src/engine/strat_engine/blockdevmgr.rs

```diff
@@ -40,15 +42,45 @@ pub fn resolve_devices(paths: &[&Path]) -> io::Result<HashSet<Device>> {
         .collect()
 }
 
+#[derive(Debug)]
+pub struct BlkDevSegment {
+    pub blockdev: Rc<RefCell<BlockDev>>,
+    pub start: Sectors,
+    pub length: Sectors,
+}
+
+impl BlkDevSegment {
+    pub fn new(blockdev: Rc<RefCell<BlockDev>>, start: Sectors, length: Sectors) -> BlkDevSegment {
+        BlkDevSegment {
+            blockdev,
+            start,
+            length,
+        }
+    }
+
+    pub fn to_segment(&self) -> Segment {
+        Segment::new(*self.blockdev.borrow().device(), self.start, self.length)
+    }
+}
+
+/// Build a Vec<Segment> from BlkDevSegments. This is useful for calls
+/// to the devicemapper library.
+pub fn convert_to_dm_segs(bsegs: &[BlkDevSegment]) -> Vec<Segment> {
+    bsegs
+        .iter()
+        .map(|bseg| bseg.to_segment())
+        .collect::<Vec<_>>()
+}
+
 
 #[derive(Debug)]
 pub struct BlockDevMgr {
-    block_devs: Vec<BlockDev>,
+    block_devs: Vec<Rc<RefCell<BlockDev>>>,
     last_update_time: Option<DateTime<Utc>>,
 }
 
 impl BlockDevMgr {
-    pub fn new(block_devs: Vec<BlockDev>) -> BlockDevMgr {
+    pub fn new(block_devs: Vec<Rc<RefCell<BlockDev>>>) -> BlockDevMgr {
```

Comment:
I think it would be nicer if the type of ```block_devs``` continued to be just ```Vec<BlockDev>``` but converted the contents to reference counted ref cells here. That way, ```initialize()``` can keep its simple return type.